File: common/src/main/java/juuxel/adorn/platform/MenuBridge.java
package juuxel.adorn.platform;

import net.minecraft.core.BlockPos;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.MenuType;
import org.jspecify.annotations.Nullable;

public interface MenuBridge {
    /**
     * Opens a menu with a pos with the opening NBT sent to the client.
     * Does nothing on the client.
     */
    void open(Player player, @Nullable MenuProvider factory, BlockPos pos);

    <M extends AbstractContainerMenu, D> MenuType<M> createType(Factory<M, D> factory, StreamCodec<? super RegistryFriendlyByteBuf, D> packetCodec);

    @FunctionalInterface
    interface Factory<M extends AbstractContainerMenu, D> {
        M create(int syncId, Inventory inventory, D data);
    }
}

File: common/src/main/kotlin/juuxel/adorn/platform/MenuBridge.kt
package juuxel.adorn.platform

import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.menu.Menu
import net.minecraft.menu.MenuType
import net.minecraft.menu.NamedMenuFactory
import net.minecraft.network.PacketByteBuf
import net.minecraft.util.math.BlockPos

interface MenuBridge {
    // This is only needed because Forge wants to be special and have a custom opening method.
    /**
     * Opens a menu with a [pos] with the opening NBT sent to the client.
     * Does nothing on the client.
     *
     * @param player  the player opening the menu
     * @param factory the menu factory
     */
    fun open(player: PlayerEntity, factory: NamedMenuFactory?, pos: BlockPos)

    fun <M : Menu> create(factory: (syncId: Int, inventory: PlayerInventory) -> M): MenuType<M>
    fun <M : Menu> create(factory: (syncId: Int, inventory: PlayerInventory, buf: PacketByteBuf) -> M): MenuType<M>
}
